xicam/plugins/__init__.py

```diff
@@ -63,19 +62,40 @@ class Filters(Enum):
     COMPLETE = auto()
 
 
+class Status(Enum):
+    LoadingQueue = auto()
+    Loading = auto()
+    InstantiateQueue = auto()
+    Instantiating = auto()
+    Success = auto()
+    FailedLoad = auto()
+    FailedInstantiate = auto()
+
+
+@dataclass
+class PluginTask:
+    type_name: str
+    name: str
+    entry_point: entrypoints.EntryPoint = field(compare=False)
+    plugin_class: Type = field(default=None, compare=False)
+    status: Status = field(default=Status.Loading, compare=False)
+
+
 class XicamPluginManager:
     def __init__(self, qt_is_safe=False):
 
         self.qt_is_safe = qt_is_safe
         self._blacklist = []
+        self._tasks = []
         self._load_queue = LifoQueue()
         self._instantiate_queue = LifoQueue()
-        self._entrypoints = {}
-        self._load_cache = {}
         self._observers = []
-        self.state = State.READY
         self.type_mapping = {}
         self.plugin_types = {}
+        self.instantiating = False
+
+        # A QRunnable-based background Worker
+        self.plugin_loader = GeneratorWorker(self._load_plugins)
```

Comment:
Is this a replacement for the threading decorator used before on `self._load_plugins`? Do you think that that decorator was potentially causing loading issues?

Reply:
No, I don't think there were any issues inherent to `xicam.core.threads`, but I wanted to rework how the thread was managed. Before there would be a new QThread generated each time `.collect_plugins` was called. There was some bookkeeping to try to make sure that two didn't run at the same time, utilitizing the `thread_key` option, but I think that doesn't work well. Anyways, I wanted to use one thread, and start its task whenever `.collect_plugins` was called. I went to check how that was done in `bluesky-widgets`, since we had some careful discussion there, and noticed that they had a nicely isolated module of worker classes, with some of the same features I'd implemented at `xicam.core.threads`. I haven't completely explored the differences between these implementations, but I can already see some advantages on either side. Their module (which actually seems to be borrowed from `napari`) uses QRunnable and a QThreadPool. The `GeneratorWorker` is expected to yield periodically to defer to a makeshift eventloop where it can change state (pause, resume, stop, ...). Since I've been recently concerned about thread cleanup, that seems like a nice advantage.

We should take a careful look at both and think about how a singular solution can be formed, either by pulling features into `xicam.core.threads` or by pushing features into `bluesky-widgets`